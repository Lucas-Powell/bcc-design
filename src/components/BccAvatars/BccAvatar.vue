<script setup lang="ts">
withDefaults(
  defineProps<{
    initials: string;
    gender?: "male" | "female" | "unknown";
    child?: boolean;
  }>(),
  {
    gender: "unknown",
    child: false,
  }
);

const colors = {
  male: ["bg-blue-100", "text-blue-900"],
  female: ["bg-violet-100", "text-violet-900"],
  unknown: ["bg-neutral-100", "text-neutral-900"],
};
const childColors = {
  male: ["bg-sky-100", "text-sky-800"],
  female: ["bg-pink-100", "text-pink-800"],
  unknown: ["bg-neutral-100", "text-neutral-800"],
};
</script>

<template>
  <div
    class="avatars-avatar rounded-full text-center font-semibold"
    :class="child ? childColors[gender] : colors[gender]"
  >
    {{ initials.slice(0, 2) }}
  </div>
</template>

<style>
.avatars-avatar {
  width: 2.285em !important;
  height: 2.275em !important;
  line-height: 2.285em !important;
  box-shadow: 0 0.071em 0.142em rgba(0, 0, 0, 0.05) !important;
}
</style>
