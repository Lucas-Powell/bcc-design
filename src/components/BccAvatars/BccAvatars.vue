<script lang="ts">
import { ref } from "vue";
import { useElementBounding, useWindowSize } from "@vueuse/core";
import { Popover, PopoverButton, PopoverPanel } from "@headlessui/vue";

export type AvatarProps = {
  list: Array<{
    name: string;
    age: number;
    gender: "male" | "female" | "unknown";
  }>;
  max?: number;
};

const el = ref();
const { top, x } = useElementBounding(el);
const { height, width } = useWindowSize();
</script>

<script setup lang="ts">
import BccAvatar from "./BccAvatar.vue";

withDefaults(defineProps<AvatarProps>(), {
  max: 4,
});
</script>

<template>
  <Popover v-slot="{ open }" class="relative">
    <PopoverButton class="relative z-0 flex items-center outline-none" ref="el">
      <BccAvatar
        v-for="(person, i) in max < list.length ? list.slice(0, max - 1) : list"
        :key="person.name"
        :initials="`${person.name.slice(0, 2)}`"
        :gender="person.gender"
        :child="person.age < 18"
        :style="{ zIndex: i * -1 }"
        class="-ml-[0.571em] shrink-0"
      />
      <BccAvatar
        v-if="max < list.length"
        :initials="`+${list.length - max + 1}`"
        :style="{ zIndex: max * -1 }"
        class="-ml-[0.571em] shrink-0"
      />
    </PopoverButton>

    <transition name="fade">
      <PopoverPanel
        v-if="open"
        class="absolute z-10"
        :class="[
          top < height / 2 ? 'top-full mt-2' : 'bottom-full mb-2',
          x > width / 2 ? 'right-0' : 'left-0',
        ]"
      >
        <div class="w-48 rounded-lg bg-white shadow-lg">
          <ul class="p-2 text-base">
            <li v-for="person in list" :key="person.name" class="flex items-center py-1">
              <BccAvatar
                :initials="`${person.name.slice(0, 2)}`"
                :gender="person.gender"
                :child="person.age < 18"
                class="mr-2 shrink-0"
              />

              <p>
                {{ person.name }} <span class="text-sm opacity-50">{{ person.age }}</span>
              </p>
            </li>
          </ul>
        </div>
      </PopoverPanel>
    </transition>
  </Popover>
</template>
