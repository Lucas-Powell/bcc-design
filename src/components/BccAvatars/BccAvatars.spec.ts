import { mount } from "@vue/test-utils";
import BccAvatars from "./BccAvatars.vue";

describe("BccAvatars", () => {
  it("mount component", async () => {
    expect(BccAvatars).toBeTruthy();

    const wrapper = mount(BccAvatars, {
      props: {
        max: 2,
        list: [
          { name: "John Doe", age: 32, gender: "m" },
          { name: "Jane Doe", age: 32, gender: "f" },
          { name: "Jack Doe", age: 12, gender: "m" },
          { name: "Jill Doe", age: 3, gender: "f" },
          { name: "Jenny Doe", age: 1, gender: "f" },
        ],
      },
    });

    expect(wrapper.html()).toMatchSnapshot();

    await wrapper.get("button").trigger("click");

    expect(wrapper.html()).toMatchSnapshot();
  });
});
