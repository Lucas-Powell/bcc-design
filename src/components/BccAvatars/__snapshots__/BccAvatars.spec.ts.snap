// Vitest Snapshot v1

exports[`BccAvatars > mount component 1`] = `
"<div data-headlessui-state=\\"\\" class=\\"relative\\"><button class=\\"relative z-0 flex items-center outline-none\\" id=\\"headlessui-popover-button-1\\" type=\\"button\\" aria-expanded=\\"false\\" data-headlessui-state=\\"\\">
    <div class=\\"avatars-avatar rounded-full text-center font-semibold -ml-[0.571em] shrink-0\\" style=\\"z-index: 0;\\">Jo</div>
    <div class=\\"avatars-avatar rounded-full text-center font-semibold bg-neutral-100 text-neutral-900 -ml-[0.571em] shrink-0\\" style=\\"z-index: -2;\\">+4</div>
  </button>
  <!---->
  <transition-stub name=\\"fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
    <!--v-if-->
  </transition-stub>
</div>"
`;

exports[`BccAvatars > mount component 2`] = `
"<div data-headlessui-state=\\"open\\" class=\\"relative\\"><button class=\\"relative z-0 flex items-center outline-none\\" id=\\"headlessui-popover-button-1\\" type=\\"button\\" aria-expanded=\\"true\\" data-headlessui-state=\\"open\\" aria-controls=\\"headlessui-popover-panel-3\\">
    <div class=\\"avatars-avatar rounded-full text-center font-semibold -ml-[0.571em] shrink-0\\" style=\\"z-index: 0; z-index: 0; z-index: 0; z-index: 0;\\">Jo</div>
    <div class=\\"avatars-avatar rounded-full text-center font-semibold bg-neutral-100 text-neutral-900 -ml-[0.571em] shrink-0\\" style=\\"z-index: -2;\\">+4</div>
  </button><button aria-hidden=\\"true\\" style=\\"position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;\\" id=\\"headlessui-focus-sentinel-2\\" data-headlessui-focus-guard=\\"true\\" type=\\"button\\"></button>
  <transition-stub name=\\"fade\\" appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
    <div class=\\"absolute z-10 top-full mt-2 left-0\\" id=\\"headlessui-popover-panel-3\\" tabindex=\\"-1\\" data-headlessui-state=\\"open\\"><button aria-hidden=\\"true\\" style=\\"position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;\\" id=\\"headlessui-focus-sentinel-before-4\\" data-headlessui-focus-guard=\\"true\\" type=\\"button\\"></button>
      <div class=\\"w-48 rounded-lg bg-white shadow-lg\\">
        <ul class=\\"p-2 text-base\\">
          <li class=\\"flex items-center py-1\\">
            <div class=\\"avatars-avatar rounded-full text-center font-semibold mr-2 shrink-0\\">Jo</div>
            <p>John Doe <span class=\\"text-sm opacity-50\\">32</span></p>
          </li>
          <li class=\\"flex items-center py-1\\">
            <div class=\\"avatars-avatar rounded-full text-center font-semibold mr-2 shrink-0\\">Ja</div>
            <p>Jane Doe <span class=\\"text-sm opacity-50\\">32</span></p>
          </li>
          <li class=\\"flex items-center py-1\\">
            <div class=\\"avatars-avatar rounded-full text-center font-semibold mr-2 shrink-0\\">Ja</div>
            <p>Jack Doe <span class=\\"text-sm opacity-50\\">12</span></p>
          </li>
          <li class=\\"flex items-center py-1\\">
            <div class=\\"avatars-avatar rounded-full text-center font-semibold mr-2 shrink-0\\">Ji</div>
            <p>Jill Doe <span class=\\"text-sm opacity-50\\">3</span></p>
          </li>
          <li class=\\"flex items-center py-1\\">
            <div class=\\"avatars-avatar rounded-full text-center font-semibold mr-2 shrink-0\\">Je</div>
            <p>Jenny Doe <span class=\\"text-sm opacity-50\\">1</span></p>
          </li>
        </ul>
      </div><button aria-hidden=\\"true\\" style=\\"position: fixed; height: 0px; padding: 0px; overflow: hidden; clip: rect(0px, 0px, 0px, 0px); white-space: nowrap; border-width: 0px;\\" id=\\"headlessui-focus-sentinel-after-5\\" data-headlessui-focus-guard=\\"true\\" type=\\"button\\"></button>
    </div>
  </transition-stub>
</div>"
`;
